lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,378 @@
+
+    """
+    def __init__(self, allbuffers):
+        """Initialise the class.
+
+        Create the global variables that allows the "filter" function
+        to save each extracted buffer into "allbuffers".
+
+        """
+        # Initialises the iterator.
+        self.i = 0
+        # Saves all the buffers into this array.
+        self.allbuffers = allbuffers
+
+    def filter(self, buffer):
+        """Defines filter function to be applied to extracted buffers.
+
+        Saves the contents of the buffer into "allbuffers" array. Therefore
+        a return value isn't required. However "generic_filter"
+        requires a return value - so use zero.
+
+        Args:
+            buffer (array):
+                Array containing neighourbood points.
+
+        Returns:
+            zero (float)
+                Blank return value required by "generic_filter".
+
+        """
+        self.allbuffers[self.i, :] = buffer
+        self.i += 1
+        return 0.0
+
+
+class LapseRate(object):
+    """Plugin to calculate the lapse rate from orography and temperature
+       cubes.
+
+    Science background:
+
+    References:
+        The method applied here is based on the method used in the 2010 paper:
+        https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/met.177
+
+    Code methodology:
+    1) Apply land/sea mask to temperature and orography datasets. Mask sea
+       points as NaN since image processing module does not recognise Numpy
+       masks.
+    2) Extracts neighbourhoods from both datasets:
+       Apply "generic_filter" image processing module to temperature data
+       to extract a neighbourhood for each single point and save this
+       neighbourhood into a larger array.
+       Repeat for orography data.
+       generic_filter mode="constant" ensures that points beyond edges of
+       dataset will be filled with "cval". (NaN in this case)
+    3) For all the stored orography neighbourhoods - take the neighbours around
+       the central point and create a mask where the height difference from
+       the central point is greater than 35m.
+    4) Loop through array of neighbourhoods and take the height and temperature
+       of all grid points and calculate the
+       temperature/height gradient = lapse rate
+    5) Constrain the lapse rate as > DALR and < -3.0*DALR
+
+    """
+
+    def __init__(self, max_height_diff=35, nbhood_radius=1,
+                 max_lapse_rate=-3*DALR, min_lapse_rate=DALR):
+        """Initialise class.
+        Args:
+            max_height_diff (float):
+                Maximum allowable height difference between the central point
+                and points in the neighbourhood over which the lapse rate will
+                be calculated (metres).
+
+            nbhood_radius (int):
+                Radius of neighbourhood around each point. Note that the
+                neighbourhood will still be a square array.
```

Comment:
If you wanted to be specific, you could say `The neighbourhood will be a square array with side length 2*nbhood_radius + 1`.